File: app/services/exchange_rate_service.rb
# frozen_string_literal: true

require_relative '../factories/exchange_rate_provider_factory'

class ExchangeRateService
  def self.get_rates(source, currencies = [])
    provider = ExchangeRateProviderFactory.create(source)
    rates = provider.request_exchange_rates

    return rates if currencies.empty?

    rates_filtered = rates[source].select { |rate| currencies.include?(rate[:currency_code]) }

    { source => rates_filtered }
  end
end

File: jobs/Backend/Task/ExchangeRateUpdater.Services/ExchangeRateService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ExchangeRateUpdater.Abstractions.Interfaces;
using ExchangeRateUpdater.Abstractions.Model;

namespace ExchangeRateUpdater.Services
{
    /// <summary>
    /// Service for retrieving exchange rates using a specified provider.
    /// </summary>
    /// <param name="provider"></param>
    public class ExchangeRateService(IExchangeRateProvider provider) : IExchangeRatesService
    {
        private readonly IExchangeRateProvider provider = provider ?? throw new ArgumentNullException(nameof(provider));

        public async Task<IReadOnlyList<ExchangeRate>> GetRates(IList<string> currencyCodes)
        {
            if (currencyCodes == null || !currencyCodes.Any())
            {
                throw new ArgumentException("Currency codes cannot be empty.");
            }

            var currencies = currencyCodes
                .Select(code => new Currency(code.Trim().ToUpperInvariant()))
                .ToArray();
            
            return (await provider.GetExchangeRates(currencies)).ToList();
        }
    }
}